sj201_revA/switch.py

```diff
@@ -0,0 +1,163 @@
+    and interrupt driven. Also note switches are
+    pulled up so the active state is actually zero.
+    """
+    # GPIO pin numbers
+    _SW_ACTION = 22
+    _SW_VOL_UP = 23
+    _SW_VOL_DOWN = 24 
+    _SW_MUTE = 25
+
+    RED = (255,0,0)
+    GREEN = (0,255,0)
+    BLUE = (0,0,255)
+    BLACK = (0,0,0)
+    WHITE = (255,255,255)
+
+    def __init__(self, debounce=100):
+        self.debounce = debounce
+        self.leds = None
+        self.volume = None
+        self.watchdog = None
+        self.watchdog_timeout = 5
+        self.active = 0
+        self.vol_increment = 0.1
+
+        # use BCM GPIO numbering
+        GPIO.setmode(GPIO.BCM)
+
+        # we need to pull up the 3 buttons
+        GPIO.setup(self._SW_ACTION, 
+            GPIO.IN, 
+            pull_up_down=GPIO.PUD_UP)
+
+        GPIO.setup(self._SW_VOL_UP, 
+            GPIO.IN, 
+            pull_up_down=GPIO.PUD_UP)
+
+        GPIO.setup(self._SW_VOL_DOWN, 
+            GPIO.IN, 
+            pull_up_down=GPIO.PUD_UP)
+
+        GPIO.setup(self._SW_MUTE, GPIO.IN)
+
+        # establish default values
+        self.SW_ACTION = GPIO.input(self._SW_ACTION)
+        self.SW_VOL_UP = GPIO.input(self._SW_VOL_UP)
+        self.SW_VOL_DOWN = GPIO.input(self._SW_VOL_DOWN)
+        self.SW_MUTE = GPIO.input(self._SW_MUTE)
+
+        # establish default handlers for each switch 
+        self.action_handler = self.handle_action
+        self.vol_up_handler = self.handle_vol_up
+        self.vol_down_handler = self.handle_vol_down
+        self.mute_handler = self.handle_mute
+
+        # attach callbacks
+        GPIO.add_event_detect(self._SW_ACTION, 
+            GPIO.BOTH, 
+            callback=self.action_handler, 
+            bouncetime=debounce)
+
+        GPIO.add_event_detect(self._SW_VOL_UP, 
+            GPIO.BOTH, 
+            callback=self.vol_up_handler, 
+            bouncetime=debounce)
+
+        GPIO.add_event_detect(self._SW_VOL_DOWN, 
+            GPIO.BOTH, 
+            callback=self.vol_down_handler, 
+            bouncetime=debounce)
+
+        GPIO.add_event_detect(self._SW_MUTE, 
+            GPIO.BOTH, 
+            callback=self.mute_handler, 
+            bouncetime=debounce)
+
+    def handle_watchdog(self):
+        # after brief delay clear the volume leds
+        self.leds.fill_leds( self.BLACK )
+        self.watchdog = None
+        
+    def handle_action(self, channel):
+        self.SW_ACTION = GPIO.input(self._SW_ACTION)
+        if self.user_action_handler is not None and self.SW_ACTION == self.active:
+            self.user_action_handler()
+
+    def handle_vol_up(self, channel):
+        self.SW_VOL_UP = GPIO.input(self._SW_VOL_UP)
+        if self.leds is not None and self.SW_VOL_UP == self.active:
+            vol = self.volume.get_hardware_volume()
+            vol += self.vol_increment
+            self.volume.set_hardware_volume(vol)
+            vol = int(vol * 10)  # note we are only using 10 of the 12 leds 
```

Comment:
why only use 10 LEDs?

Reply:
Ah, good question. First it is much easier to just use a 1/10 step for volume to led mapping but additionally I suspect at least 2 LEDs will be used for other purposes down stream (listening indicator, etc) so rather than refactoring and changing code I decided to use just 10. If it turns out we will not be needing LEDs for anything but volume indicator I can use all 12.